crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+    pub height: u32,
+}
+
+/// A block header obtained from `bitcoind`.
+pub struct EmittedHeader {
+    /// The actual block header.
+    pub header: Header,
+    /// The height of the block header.
+    pub height: u32,
+}
+
+/// A structure that emits data sourced from [`bitcoincore_rpc::Client`].
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub struct Emitter<'c, C> {
+    client: &'c C,
+    start_height: u32,
+
+    emitted: BTreeMap<u32, BlockHash>,
+    last: Option<bitcoincore_rpc_json::GetBlockResult>,
+}
+
+impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Constructs a new [`Emitter`] with the provided [`bitcoincore_rpc::Client`].
+    ///
+    /// `start_height` is the block height to start emitting blocks from.
+    pub fn new(client: &'c C, start_height: u32) -> Self {
+        Self {
+            client,
+            start_height,
+            emitted: BTreeMap::new(),
+            last: None,
+        }
+    }
+
+    /// Emit all mempool transactions.
+    pub fn mempool(&self) -> impl Iterator<Item = Result<MempoolTx, bitcoincore_rpc::Error>> + '_ {
+        let res = self.client.get_raw_mempool();
+
+        let tx_iter = res.as_ref().unwrap_or(&Vec::new()).clone().into_iter().map(
+            |txid| -> Result<MempoolTx, bitcoincore_rpc::Error> {
+                let tx = self.client.get_raw_transaction(&txid, None)?;
+                let time = self.client.get_mempool_entry(&txid)?.time;
+                Ok(MempoolTx { tx, time })
+            },
+        );
+        // iterator will be non-empty if get_raw_mempool returned an error
+        let err_iter = res.err().into_iter().map(Err);
+
+        err_iter.chain(tx_iter)
+    }
+
+    /// Emit the next block header (if any).
+    pub fn next_header(&mut self) -> Result<Option<EmittedHeader>, bitcoincore_rpc::Error> {
+        let poll_res = poll(self, |hash| self.client.get_block_header(hash))?;
+        Ok(poll_res.map(|(height, header)| EmittedHeader { header, height }))
+    }
+
+    /// Emit the next block (if any).
+    pub fn next_block(&mut self) -> Result<Option<EmittedBlock>, bitcoincore_rpc::Error> {
+        let poll_res = poll(self, |hash| self.client.get_block(hash))?;
+        Ok(poll_res.map(|(height, block)| EmittedBlock { block, height }))
+    }
+
+    /// Transforms `self` into an iterator of [`Emission`]s.
+    ///
+    /// Refer to [module-level documentation] for more.
+    ///
+    /// [module-level documentation]: crate
+    pub fn into_iterator<B>(self) -> Iter<'c, C, B> {
+        Iter {
+            emitter: self,
+            last_emission_was_mempool: false,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// This is the [`Iterator::Item`] of [`Iter`]. This can either represent a block/header or the set
+/// of mempool transactions.
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub enum Emission<B> {
```

Comment:
Why do these different data requests (mempool, next block) needed to be handled in the same "stream" of events. Looking at the caller code it doesn't seem to simplify things, they are always matched on and handled differently (this could be because of the difference in vision I have kept pushing you into a corner where this was the case).

Furthermore, A user is likely to want to check for the next block vs dump the mempool at different times. E.g. dump the mempool every 30s but check for next block every 5s (actually it would be nice if we did this in our example). I strongly believe the caller should make a conscious decision about when to do this. It is not a natural decision to me to dump the mempool after you've finished looking for new transactions (and then continuing to dump it at every second call to `next` which feels even weirder).

Reply:
If we scan the mempool for relevant txs before scanning the latest tip, we may end up excluding mempool transactions that send from our UTXO pool (as we need the prev txs to determine relevancy).

This may be problematic if we intend to have non-duplicate tx dumps from the mempool (as suggested in https://github.com/bitcoindevkit/bdk/pull/1041#discussion_r1315406687). We will end up excluding an actually-relevant mempool transaction which sends from our wallet.

If this tx takes a while to confirm, we can be fully synced up to the tip and not see it in the wallet until it is confirmed. The wallet may accidentally rbf the tx if the tx stays unconfirmed.

I think it is good policy to scan the mempool only after scanning the latest tip.